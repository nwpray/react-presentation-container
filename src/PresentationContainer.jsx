import React from 'react';
import isFunction from 'lodash/isFunction';

const pipe = (fncs, initValue) => fncs.reduce((prevValue, fn) => fn(prevValue), initValue);

export default ({
  component: Component,
  filterProps,
  bindMembers = [],
  controller: Controller,
  middleware = []
}) => {
  const getContainer = () => {
    if (isFunction(Controller) && !!Controller.prototype.isReactComponent) {
      /* eslint-disable react/no-this-in-sfc */
      /* eslint-disable no-param-reassign */
      Controller.prototype.render = function controllerRender() {
        const boundMembers = bindMembers.reduce(
          (prevMembers, key) => ({ ...prevMembers, [key]: this[key] }),
          {}
        );

        return (
          <Component
            {...(filterProps ? filterProps(this.props) : this.props)}
            controller={{ ...boundMembers, state: this.state }}
          />
        );
      };
      /* eslint-enable */

      return Controller;
    }

    if (isFunction(Controller)) {
      if (bindMembers) {
        // eslint-disable-next-line no-console
        console.warn(
          `PresentationContainer warning: using bindMembers with a functional controller is not supported (${Controller.prototype.constructor.name})`
        );
      }

      return function FunctionalContainer(props, ...restOfArgs) {
        const state = Controller(props, ...restOfArgs);

        return <Component {...(filterProps ? filterProps(props) : props)} controller={state} />;
      };
    }

    return null;
  };

  const ContainerComponent = getContainer();

  return ContainerComponent ? pipe(middleware, ContainerComponent) : null;
};
