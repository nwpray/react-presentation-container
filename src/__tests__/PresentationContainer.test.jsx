import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';

import PresentationContainer from '../PresentationContainer';

const { describe, expect, it, jest } = global;

describe('PresentationContainer() (class controller)', () => {
  it('renders', () => {
    const MockChild = () => {
      return <h1>Working</h1>;
    };

    const WrappedComponent = PresentationContainer({
      component: MockChild,
      controller: class MockChildController extends React.Component {}
    });

    const { container } = render(<WrappedComponent />);

    expect(container).toMatchSnapshot();
  });
});

describe('PresentationContainer() (function controller)', () => {
  it('renders', () => {
    const MockChild = () => {
      return <h1>Working</h1>;
    };

    const WrappedComponent = PresentationContainer({
      component: MockChild,
      controller: function MockChildController() {}
    });

    const { container } = render(<WrappedComponent />);

    expect(container).toMatchSnapshot();
  });

  it('filters props', () => {
    const MockChild = () => {
      return <h1>Working</h1>;
    };

    const WrappedComponent = PresentationContainer({
      component: MockChild,
      filterProps: ({ propA }) => ({ propA }),
      controller: function MockChildController() {}
    });

    const { container } = render(<WrappedComponent propA="propA" propB="propB" propC="propC" />);
  });
});

// test('form submits two answers', () => {
//   const allQuestions = ['q1', 'q2'];
//   const mockFn = jest.fn();

//   const { getAllByA11yLabel, getByText } = render(
//     <QuestionsBoard questions={allQuestions} onSubmit={mockFn} />
//   );

//   const answerInputs = getAllByA11yLabel('answer input');

//   fireEvent.changeText(answerInputs[0], 'a1');
//   fireEvent.changeText(answerInputs[1], 'a2');
//   fireEvent.press(getByText('Submit'));

//   expect(mockFn).toBeCalledWith({
//     '1': { q: 'q1', a: 'a1' },
//     '2': { q: 'q2', a: 'a2' }
//   });
// });
