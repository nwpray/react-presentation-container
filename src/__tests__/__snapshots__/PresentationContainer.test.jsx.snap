// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PresentationContainer() renders with class controller 1`] = `
<div>
  <h1>
    Working
  </h1>
</div>
`;

exports[`PresentationContainer() renders with function controller 1`] = `
<div>
  <h1>
    Working
  </h1>
</div>
`;
